### PR TITLE
Topic/repl customization

### PR DESCRIPTION
**Pull request by iracic**
Few low hanging fruit changes to customize scala REPL. 
Please, let me know if anything should be improved.

### CODE REVIEW

**Comment by adriaanm** [comment]
Welcome & thank you, @iracic! Please [sign the Scala CLA](http://typesafe.com/contribute/cla/scala).

File: src/compiler/scala/tools/nsc/Properties.scala [hunk 1]
```diff
@@ -13,7 +13,7 @@ object Properties extends scala.util.PropertiesTrait {
 
   // settings based on jar properties, falling back to System prefixed by "scala."
   def residentPromptString = scalaPropOrElse("resident.prompt", "\nnsc> ")
-  def shellPromptString    = scalaPropOrElse("shell.prompt", "\nscala> ")
+  def shellPromptString    = "\n" + Properties.propOrElse("scala.repl.PS1", "scala> ")
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/Properties.scala]: Wouldn't this break existing customizers? I don't think we can afford to do that, sorry.

Comment (iracic) [inline src/compiler/scala/tools/nsc/Properties.scala]: You are right, I was too fast here. 
I propose to keep this one and also try to change namespace as "scala.shell." on other properties in the same way. 
I will rework/rename these and try again (should I go with new PR) ?

Comment (iracic) [inline src/compiler/scala/tools/nsc/Properties.scala]: I have created another branch topic/repl-customization--2nd-try. 
Let me know your opinion about it. 
Thank you.

File: src/repl/scala/tools/nsc/interpreter/session/FileBackedHistory.scala [hunk 2]
```diff
@@ -80,5 +80,5 @@ object FileBackedHistory {
   import Properties.userHome
 
   def defaultFileName = ".scala_history"
-  def defaultFile: File = File(Path(userHome) / defaultFileName)
+  def defaultFile: File = File( Path( Properties.propOrElse( "scala.repl.histfile", userHome +"/"+ defaultFileName) ) )
```
Comment (gourlaysama) [inline src/repl/scala/tools/nsc/interpreter/session/FileBackedHistory.scala]: Using `+ "/" +` to create the path fails to account for Windows users and their backslashes. That's what the `/` method on `Path` does. Maybe something like:

``` scala
File(Properties.propOrNone("scala.repl.histfile").fold(Path(userHome) / defaultFileName))(Path(_)))
```

File: src/compiler/scala/tools/nsc/Properties.scala (hunk 1 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/Properties.scala]: Ok, please push --force to the same branch so that this PR is updated.

Comment (mpociecha) [inline src/compiler/scala/tools/nsc/Properties.scala]: Maybe with squash?

File: src/repl/scala/tools/nsc/interpreter/ILoop.scala [hunk 3]
```diff
@@ -915,7 +916,8 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
       !intp.reporter.hasErrors
     }
     loadFiles(settings)
-    printWelcome()
+    if( ! Properties.propOrFalse( "scala.repl.disableWelcomeMessage" ) ) 
+      printWelcome()
```
Comment (som-snytt) [inline src/repl/scala/tools/nsc/interpreter/ILoop.scala:933]: [Pasting my comment onto the diff.]

`ReplProps` collects various props that you can never remember the name of. Maybe time for `~/.scalarc`.

Would also be neat if the string property were a dollar quote, naturally with `$intp` in scope. If there are no dollars (a constant), spit it out immediately; otherwise, compile it and echo it in `loopPostInit`. An empty string means no banner. That's just an idea.

https://github.com/scala/scala/blob/2.11.x/src/repl/scala/tools/nsc/interpreter/ReplProps.scala

Also, the surrounding source is not quite so space happy.

File: src/repl/scala/tools/nsc/interpreter/ILoop.scala [hunk 4]
```diff
@@ -695,7 +695,8 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
   def verbosity() = {
     val old = intp.printResults
     intp.printResults = !old
-    echo("Switched " + (if (old) "off" else "on") + " result printing.")
+    if( ! Properties.propOrFalse( "scala.repl.silentVerbositySwitch" ) )
+      echo("Switched " + (if (old) "off" else "on") + " result printing.")
```
Comment (som-snytt) [inline src/repl/scala/tools/nsc/interpreter/ILoop.scala:699]: The property is very specific. I would have to look at the source to guess what it does.

However, `:silent -silent` is kind of cool in a weird way, to achieve the same unprinted result. Abbrev `:si -s`.

Or a `ReplProp` called `scala.repl.mute` or `warn` by analogy to log levels and `scala.repl.info`.

File: src/repl/scala/tools/nsc/interpreter/session/FileBackedHistory.scala (hunk 2 above)
Comment (som-snytt) [inline src/repl/scala/tools/nsc/interpreter/session/FileBackedHistory.scala]: I think slash is OK internally on Windows.

Less cleverly, `Option(propOrNull) map (File(_)) getOrElse File(userHome, default)`. Or whatever.

File: src/compiler/scala/tools/nsc/Properties.scala (hunk 1 above)
Comment (iracic) [inline src/compiler/scala/tools/nsc/Properties.scala]: Ok, will do forced push. 
Adriaan, do you prefer to squash to one change ?

File: src/repl/scala/tools/nsc/interpreter/ILoop.scala (hunk 3 above)
Comment (iracic) [inline src/repl/scala/tools/nsc/interpreter/ILoop.scala:933]: Thank you for comment. 
Haven't spotted ReplProps before. 
It is a bit unfortunate that configs are in different namespaces "scala.shell" vs "scala.repl". 

Things about string evaluation are maybe too strong for me in this stage... :-)

File: src/repl/scala/tools/nsc/interpreter/ILoop.scala (hunk 4 above)
Comment (iracic) [inline src/repl/scala/tools/nsc/interpreter/ILoop.scala:699]: Agreed, it will be easy to forget. 
Not sure alternatives are much clearer.

File: src/repl/scala/tools/nsc/interpreter/ILoop.scala (hunk 3 above)
Comment (iracic) [inline src/repl/scala/tools/nsc/interpreter/ILoop.scala:933]: Not sure I understood comment about space happiness :-|

File: src/repl/scala/tools/nsc/interpreter/session/FileBackedHistory.scala (hunk 2 above)
Comment (iracic) [inline src/repl/scala/tools/nsc/interpreter/session/FileBackedHistory.scala]: I did quick test with slash on Windows and it seems to be OK.

Comment (gourlaysama) [inline src/repl/scala/tools/nsc/interpreter/session/FileBackedHistory.scala]: Ok, all modern Windows handle it fine, and the java File API normalizes everything that comes through the public API anyway, so it all works (I can't believe I am complimenting the Java 6 File API). Just ignore my comments :-)

**Comment by adriaanm** [comment]
Yes, please squash to one commit, and take a look at the [guidelines around formatting your commit message](https://github.com/scala/scala#git-hygiene) (short title that can be included in release notes, description in body, reference to jira ticket if any,...). Here's a rough suggestion for the title: "Configurable REPL: prompt, verbosity, history path".

**Comment by iracic** [comment]
Squashed. Thank you.

**Comment by soc** [comment]
There are also various places where the change introduced odd formatting, this should be fixed, too.

**Comment by som-snytt** [comment]
I think my previous comments about formatting and using `ReplProps` are still relevant.

In anticipation, I have prepared a StackOverflow answer to "How do I set the history file for REPL?"

**Comment by iracic** [comment]
Tried to reformat code and put properties in ReplProps

**Comment by iracic** [comment]
Above failure seem like infrastructure errors. I do not find link between them and code changes...

**Comment by mpociecha** [comment]
You're right, it's not related to your changes. We're looking at these problems.

**Comment by iracic** [comment]
@mpociecha Thank you

**Comment by mpociecha** [comment]
@iracic Simply I'm a culprit.

**Comment by iracic** [comment]
@mpociecha Shame on you :-)

File: src/repl/scala/tools/nsc/interpreter/ReplProps.scala [hunk 5]
```diff
@@ -32,6 +32,12 @@ class ReplProps {
   val powerInitCode   = Prop[JFile]("scala.repl.power.initcode")
   val powerBanner     = Prop[JFile]("scala.repl.power.banner")
 
+  val prompt = Prop[String]("scala.shell.prompt")
+  val prompt2 = Prop[String]("scala.shell.prompt2")
```
Comment (lrytz) [inline src/repl/scala/tools/nsc/interpreter/ReplProps.scala:36]: `prompt2` should be renamed to something more descriptive

File: src/repl/scala/tools/nsc/interpreter/ReplProps.scala [hunk 6]
```diff
@@ -32,6 +32,12 @@ class ReplProps {
   val powerInitCode   = Prop[JFile]("scala.repl.power.initcode")
   val powerBanner     = Prop[JFile]("scala.repl.power.banner")
 
+  val prompt = Prop[String]("scala.shell.prompt")
```
Comment (lrytz) [inline src/repl/scala/tools/nsc/interpreter/ReplProps.scala:35]: How is that related to `"shell.prompt"` that we already have here? https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/Properties.scala#L16

File: src/repl/scala/tools/nsc/interpreter/ReplProps.scala (hunk 5 above)
Comment (iracic) [inline src/repl/scala/tools/nsc/interpreter/ReplProps.scala:36]: I thought that "suffix" 2 make sense (similar to PS2 in Unix shell)
Let me know what do you propose ?

File: src/repl/scala/tools/nsc/interpreter/ReplProps.scala (hunk 6 above)
Comment (iracic) [inline src/repl/scala/tools/nsc/interpreter/ReplProps.scala:35]: Excellent question. Maybe I do not see it correctly, I think, prompt should be rather in Repl than in Properties. And as Properties is object, injecting repl instance is not straitforward.

**Comment by SethTisue** [comment]
superseded by #4576
